### PR TITLE
Modernizing AngelScript

### PR DESCRIPTION
**Pull request by only-a-ptr**
AngelScript in RoR needs update: http://www.angelcode.com/angelscript/downloads.html

* We're using version 2.22.1 (dated 2011)
* The current stable is 2.31.2 (dated december 2016)

Before the update itself, I want to audit the existing scripting subsystem and remove what we don't really need. I already found some garbage, see commits below.

See also:
* https://github.com/RigsOfRods/rigs-of-rods/issues/1372 - Current addons break with VS2017
* https://github.com/RigsOfRods/rigs-of-rods/pull/999 - Old PR which updates angelscript, useful as reference.
* https://sourceforge.net/u/neorej16/rigsofrods/commit_browser - Old AngelScript update by Neorej16

### CODE REVIEW

**Comment by AnotherFoxGuy** [comment]
Also related: #441

**Comment by only-a-ptr** [comment]
Status: build and starts up for me. However, when I try loading Auriga map -> :boom:

**Comment by only-a-ptr** [comment]
Status: Auriga loads fine for me, but the time-attack race in there doesn't work.

**Comment by only-a-ptr** [comment]
Status: Auriga loads fine for me, but the time-attack race in there doesn't work.

**Comment by only-a-ptr** [comment]
Status: I've just raced in Auriga with latest AngelScript! Give me some more maps/scenarios to test!

**Comment by tritonas00** [comment]
> I've just raced in Auriga with latest AngelScript!

https://www.youtube.com/watch?v=4QPdn84DuGc

:smile:

**Comment by AnotherFoxGuy** [comment]
> @only-a-ptr Give me some more maps/scenarios to test!

http://anotherfoxguy.com/Files/Benchmark.zip

**Comment by tritonas00** [comment]
I'm getting: https://pastebin.com/ktshPxm2

File: tools/travis/linux/dependencies.sh [hunk 1]
```diff
@@ -82,15 +82,15 @@ make install
...
+wget http://www.angelcode.com/angelscript/sdk/files/angelscript_2.31.2.zip
 unzip -qq angelscript_*.zip
 cd sdk/angelscript/projects/cmake
 cmake -DCMAKE_INSTALL_PREFIX=$DEPS_INSTALL_DIR \
 -DCMAKE_BUILD_TYPE:STRING=Release \
 -DCMAKE_CXX_FLAGS="-w -O0 -pipe" \
 .
 make -s -j2 
-cp -r ../../lib/libAngelscript.a $DEPS_INSTALL_DIR/lib/libangelscript.a
+cp -r ../../lib/libangelscript.a $DEPS_INSTALL_DIR/lib/libangelscript.a
```
Comment (AnotherFoxGuy) [inline tools/travis/linux/dependencies.sh]: Try ```cp -r libangelscript.a $DEPS_INSTALL_DIR/lib/libangelscript.a```

**Comment by only-a-ptr** [comment]
Eh, nothing's working.
I updated Travis dependencies, but it looks like angelscript's build output layout changed or something. It's failing with "no such file or directory". @AnotherFoxGuy I'll try it right away, thanks.
@tritonas00

**Comment by tritonas00** [comment]
yep, now travis proceeded, same AddWaypoint error

**Comment by tritonas00** [comment]
Compiled fine! Now, tried with benchmark map: https://pastebin.com/A0g8LTeE

**Comment by only-a-ptr** [comment]
@tritonas00 Apparently the benchmark map is out of date. The missing function `navigateTo()` was removed by @AnotherFoxGuy himself in a26bd842777c48acacdf624abecf6ea9ed435ab6

**Comment by tritonas00** [comment]
About the Futurotrack crash: https://pastebin.com/wDT1zdBb

**Comment by only-a-ptr** [comment]
@tritonas00 Oops, my bad, that line is bound to crash whenever the searched function doesn't exist. Fix pending...

**Comment by tritonas00** [comment]
compiles and crashes gone :thumbsup:

**Comment by AnotherFoxGuy** [comment]
@only-a-ptr @tritonas00 The link for the benchmark terrain was old, here is a new one: https://www.dropbox.com/s/lgkjj2zurpjhab5/Benchmark.zip?dl=0

**Comment by tritonas00** [comment]
Looks good!
2017-09-03-200517_1366x768_scrot

Why clang keeps failing?

**Comment by tritonas00** [comment]
GPU Benchmark from Benchmark.zip segfaults, gdb output coming.

CPU Benchmark works fine.

**Comment by only-a-ptr** [comment]
@tritonas00 

> Why clang keeps failing?

It's weird. Looks like Clang has somehow (is it even possible?)  disabled support for C++11, at least when building "angelscript_addons". It complains about lack of http://www.cplusplus.com/reference/type_traits/is_trivially_default_constructible/, although the respective header is definitely included, otherwise the offending code would get skipped.

**Comment by only-a-ptr** [comment]
@tritonas00 

> GameScript.cpp line 372: https://github.com/RigsOfRods/rigs-of-rods/pull/1373/files#diff-6f778f681dbfde564fd26c6d79e9ddb7R372

Your excellent testing report pointed me straight to the bug - another unchecked handler-function lookup. I didn't have time to test the fix, hopefully it works.

**Comment by tritonas00** [comment]
Crash fixed! But i get some errors: https://pastebin.com/UhzKMRSP

**Comment by only-a-ptr** [comment]
@tritonas00 🤣 That's an extra error report I added - the error passed silently before. I had no idea it's going to be so noisy.

@AnotherFoxGuy What is the 'e' handler supposed to be?

**Comment by AnotherFoxGuy** [comment]
> @only-a-ptr  What is the 'e' handler supposed to be?

Nothing, the handler isn't used since it doesn't have a collision box

**Comment by only-a-ptr** [comment]
Eh, on second check, I actually introduced a bug in the last commit... the object should be spawned regardless of the handler being found 😊 

That, however, sucks because this way I can't report an error where the user actually wants to use a handler and it's not found by mistake. A solution would be constitute that passing empty string as param `eventhandler` to `spawnObject()`  means "don't search for handler".

**Comment by tritonas00** [comment]
Now works fine, trees loaded and my gpu really sucks :laughing:

**Comment by only-a-ptr** [comment]
It looks like this PR is ready to merge.

The only remaining issue is the Travis/clang build failing due to https://github.com/RigsOfRods/rigs-of-rods/pull/1373#issuecomment-326827928 @AnotherFoxGuy Can you update the build machine once this PR is merged?

**Comment by AnotherFoxGuy** [comment]
> @only-a-ptr Can you update the build machine once this PR is merged?

Yes, I was planning on updating the Travis build anyway because of #1364

**Comment by only-a-ptr** [comment]
Yay, this PR is what I call a success. 🎉 
Thanks everybody.

**Pull request merged**